src/map/ai/controllers/spirit_controller.cpp

```diff
@@ -0,0 +1,593 @@
+
+#include "../../../common/utils.h"
+#include "../../entities/petentity.h"
+#include "../../status_effect_container.h"
+#include "../../utils/battleutils.h"
+#include "../../utils/petutils.h"
+#include "../../utils/zoneutils.h"
+#include "../ai_container.h"
+
+CSpiritController::CSpiritController(CPetEntity* PSpirit)
+: CPetController(PSpirit)
+, PSpirit(PSpirit)
+{
+}
+
+void CSpiritController::Tick(time_point tick)
+{
+    TracyZoneScoped;
+
+    if ((PSpirit->PMaster == nullptr || PSpirit->PMaster->isDead()) && PSpirit->isAlive())
+    {
+        PSpirit->Die();
+        return;
+    }
+
+    if (PSpirit->shouldDespawn(tick))
+    {
+        petutils::DespawnPet(PSpirit->PMaster);
+        return;
+    }
+
+    if (!PSpirit->m_setup)
+    {
+        setMagicCooldowns();
+
+        if (PSpirit->m_PetID == PETID::PETID_LIGHTSPIRIT)
+        {
+            LoadLightSpiritSpellList();
+        }
+
+        PSpirit->m_LastMagicTime = tick;
+        PSpirit->m_setup         = true;
+    }
+
+    if (PSpirit->PAI->IsEngaged())
+    {
+        PTarget = static_cast<CBattleEntity*>(PSpirit->GetEntity(PSpirit->GetBattleTargetID()));
+        if (CPetController::TryDeaggro())
+        {
+            Disengage();
+            return;
+        }
+
+        PSpirit->PAI->EventHandler.triggerListener("COMBAT_TICK", CLuaBaseEntity(PSpirit));
+        luautils::OnMobFight(PSpirit, PTarget);
+        if (TrySpellcast(tick))
+        {
+            PSpirit->m_LastMagicTime = tick;
+            return;
+        }
+
+        CPetController::Move();
+    }
+    else
+    {
+        // if pet can't follow then don't
+        if (!PSpirit->PAI->CanFollowPath())
+        {
+            return;
+        }
+        if (TryIdleSpellcast(tick))
+        {
+            PSpirit->m_LastMagicTime = tick;
+            return;
+        }
+        else
+        {
+            CPetController::DoRoamTick(tick);
+        }
+    }
+
+    CMobController::Tick(tick);
+}
+
+void CSpiritController::setMagicCooldowns()
+{
+    uint32 castTime = ((48000 + (GetSMNSkillReduction() / 3)) + GetDayWeatherBonus());
+
+    if (PSpirit->PMaster->StatusEffectContainer->HasStatusEffect(EFFECT_ASTRAL_FLOW))
+    {
+        castTime -= 5000;
+    }
+
+    CItemEquipment* legSlotItem = static_cast<CCharEntity*>(PSpirit->PMaster)->getEquip(SLOT_LEGS);
+
+    // Summoner's Spats & Summoner's Spats +1
+    if (legSlotItem && (legSlotItem->getID() == 15131 || legSlotItem->getID() == 15594))
+    {
+        castTime -= 5000;
+    }
+
+    PSpirit->m_magicCooldown = std::chrono::milliseconds(castTime);
+}
+
+bool CSpiritController::TrySpellcast(time_point tick)
+{
+    if (!PSpirit->PMaster || PSpirit->PMaster->objtype != TYPE_PC ||
+        m_Tick <= PSpirit->m_LastMagicTime + PSpirit->m_magicCooldown ||
+        PSpirit->StatusEffectContainer->HasPreventActionEffect() ||
+        PSpirit->StatusEffectContainer->HasStatusEffect(EFFECT_SILENCE))
+    {
+        return false;
+    }
+    else
```

Comment:
Don't need this else, right? And then can outdent the block below

Reply:
Removed, also removed from the other function with a similar asetup.